Crash when generating configuration for an imported object that has sensitive attributes

### Terraform Version

```shell
1.8.0
```


### Terraform Configuration Files

```terraform
terraform {
    required_version = ">= 1.8.0"
    required_providers {
      azurerm = {
        source  = "hashicorp/azurerm"
        version = ">=3.99.0"
      }
  }
}

provider "azurerm" {
  skip_provider_registration = true
  features {
    
  }
}

import {
  to = azurerm_windows_function_app.demo-func-app
  id = "/subscriptions/xxxxxxxx/resourceGroups/demo-func-rg/providers/Microsoft.Web/sites/demo-func-mt"
}

#terraform plan '-generate-config-out=generated_resources.tf'
```


### Debug Output

```
!!!!!!!!!!!!!!!!!!!!!!!!!!! TERRAFORM CRASH !!!!!!!!!!!!!!!!!!!!!!!!!!!!

Terraform crashed! This is always indicative of a bug within Terraform.
Please report the crash with Terraform[1] so that we can fix this.

When reporting bugs, please include your terraform version, the stack trace
shown below, and any additional information which may help replicate the issue.

[1]: https://github.com/hashicorp/terraform/issues

!!!!!!!!!!!!!!!!!!!!!!!!!!! TERRAFORM CRASH !!!!!!!!!!!!!!!!!!!!!!!!!!!!

panic: value is marked, so must be unmarked first
goroutine 75 [running]:
runtime/debug.Stack()
        /Users/runner/hostedtoolcache/go/1.22.1/x64/src/runtime/debug/stack.go:24 +0x64
github.com/hashicorp/terraform/internal/logging.PanicHandler()
        /Users/runner/work/terraform/terraform/internal/logging/panic.go:84 +0x198
panic({0x10378cb40?, 0x103e31460?})
        /Users/runner/hostedtoolcache/go/1.22.1/x64/src/runtime/panic.go:770 +0x124
github.com/zclconf/go-cty/cty.Value.assertUnmarked(...)
        /Users/runner/go/pkg/mod/github.com/zclconf/go-cty@v1.14.3/cty/marks.go:141
github.com/zclconf/go-cty/cty.Value.ElementIterator({{{0x103e74108?, 0x14000818f40?}}, {0x103a60180?, 0x14001854fd8?}})
        /Users/runner/go/pkg/mod/github.com/zclconf/go-cty@v1.14.3/cty/value_ops.go:1191 +0x54
github.com/hashicorp/terraform/internal/genconfig.omitUnknowns({{{0x103e74108, 0x14000818f40}}, {0x103a60180, 0x14001854fd8}})
        /Users/runner/work/terraform/terraform/internal/genconfig/generate_config.go:589 +0xd8
github.com/hashicorp/terraform/internal/genconfig.omitUnknowns({{{0x103e73ff0, 0x140009807c0}}, {0x1038c65c0, 0x14001ec0e70}})
        /Users/runner/work/terraform/terraform/internal/genconfig/generate_config.go:611 +0x20c
github.com/hashicorp/terraform/internal/genconfig.GenerateResourceContents({{}, {0x1059345e0, 0x0, 0x0}, {{}, {{}, 0x4d, {0x140001103a0, 0x1c}, {0x14000b981e0, ...}}, ...}}, ...)
        /Users/runner/work/terraform/terraform/internal/genconfig/generate_config.go:41 +0x38c
github.com/hashicorp/terraform/internal/terraform.(*NodePlannableResourceInstance).generateHCLStringAttributes(0x14000502e70, {{}, {0x1059345e0, 0x0, 0x0}, {{}, {{}, 0x4d, {0x140001103a0, 0x1c}, ...}, ...}}, ...)
        /Users/runner/work/terraform/terraform/internal/terraform/node_resource_plan_instance.go:688 +0x154
github.com/hashicorp/terraform/internal/terraform.(*NodePlannableResourceInstance).importState(0x14000502e70, {0x103e93c68, 0x140005585a0}, {{}, {0x1059345e0, 0x0, 0x0}, {{}, {{}, 0x4d, ...}, ...}}, ...)
        /Users/runner/work/terraform/terraform/internal/terraform/node_resource_plan_instance.go:624 +0x10c4
github.com/hashicorp/terraform/internal/terraform.(*NodePlannableResourceInstance).managedResourceExecute(0x14000502e70, {0x103e93c68, 0x140005585a0})
        /Users/runner/work/terraform/terraform/internal/terraform/node_resource_plan_instance.go:169 +0x3d0
github.com/hashicorp/terraform/internal/terraform.(*NodePlannableResourceInstance).Execute(0x18?, {0x103e93c68?, 0x140005585a0?}, 0xb8?)
        /Users/runner/work/terraform/terraform/internal/terraform/node_resource_plan_instance.go:75 +0x7c
github.com/hashicorp/terraform/internal/terraform.(*ContextGraphWalker).Execute(0x14001460280, {0x103e93c68, 0x140005585a0}, {0x12d3310b8, 0x14000502e70})
        /Users/runner/work/terraform/terraform/internal/terraform/graph_walk_context.go:153 +0xa8
github.com/hashicorp/terraform/internal/terraform.(*Graph).walk.func1({0x103df9560, 0x14000502e70})
        /Users/runner/work/terraform/terraform/internal/terraform/graph.go:112 +0x478
github.com/hashicorp/terraform/internal/dag.(*Walker).walkVertex(0x1400158de60, {0x103df9560, 0x14000502e70}, 0x140015d6b00)
        /Users/runner/work/terraform/terraform/internal/dag/walk.go:384 +0x2a8
created by github.com/hashicorp/terraform/internal/dag.(*Walker).Update in goroutine 63
        /Users/runner/work/terraform/terraform/internal/dag/walk.go:307 +0xc30
```

### Expected Behavior

.

### Actual Behavior

.

### Steps to Reproduce

.

### Additional Context

_No response_

### References

_No response_

--- a/internal/genconfig/generate_config.go
+++ b/internal/genconfig/generate_config.go
@@ -38,7 +38,6 @@ func GenerateResourceContents(addr addrs.AbsResourceInstance,
 		buf.WriteString(fmt.Sprintf("provider = %s\n", pc.StringCompact()))
 	}
 
-	stateVal = omitUnknowns(stateVal)
 	if stateVal.RawEquals(cty.NilVal) {
 		diags = diags.Append(writeConfigAttributes(addr, &buf, schema.Attributes, 2))
 		diags = diags.Append(writeConfigBlocks(addr, &buf, schema.BlockTypes, 2))
@@ -151,11 +150,17 @@ func writeConfigAttributesFromExisting(addr addrs.AbsResourceInstance, buf *stri
 				val = attrS.EmptyValue()
 			}
 			if val.Type() == cty.String {
+				// Before we inspect the string, take off any marks.
+				unmarked, marks := val.Unmark()
+
 				// SHAMELESS HACK: If we have "" for an optional value, assume
 				// it is actually null, due to the legacy SDK.
-				if !val.IsNull() && attrS.Optional && len(val.AsString()) == 0 {
-					val = attrS.EmptyValue()
+				if !unmarked.IsNull() && attrS.Optional && len(unmarked.AsString()) == 0 {
+					unmarked = attrS.EmptyValue()
 				}
+
+				// Before we carry on, add the marks back.
+				val = unmarked.WithMarks(marks)
 			}
 			if attrS.Sensitive || val.IsMarked() {
 				buf.WriteString("null # sensitive")
@@ -566,60 +571,4 @@ func ctyCollectionValues(val cty.Value) []cty.Value {
 	}
 
 	return ret
-}
-
-// omitUnknowns recursively walks the src cty.Value and returns a new cty.Value,
-// omitting any unknowns.
-//
-// The result also normalizes some types: all sequence types are turned into
-// tuple types and all mapping types are converted to object types, since we
-// assume the result of this is just going to be serialized as JSON (and thus
-// lose those distinctions) anyway.
-func omitUnknowns(val cty.Value) cty.Value {
-	ty := val.Type()
-	switch {
-	case val.IsNull():
-		return val
-	case !val.IsKnown():
-		return cty.NilVal
-	case ty.IsPrimitiveType():
-		return val
-	case ty.IsListType() || ty.IsTupleType() || ty.IsSetType():
-		var vals []cty.Value
-		it := val.ElementIterator()
-		for it.Next() {
-			_, v := it.Element()
-			newVal := omitUnknowns(v)
-			if newVal != cty.NilVal {
-				vals = append(vals, newVal)
-			} else if newVal == cty.NilVal {
-				// element order is how we correlate unknownness, so we must
-				// replace unknowns with nulls
-				vals = append(vals, cty.NullVal(v.Type()))
-			}
-		}
-		// We use tuple types always here, because the work we did above
-		// may have caused the individual elements to have different types,
-		// and we're doing this work to produce JSON anyway and JSON marshalling
-		// represents all of these sequence types as an array.
-		return cty.TupleVal(vals)
-	case ty.IsMapType() || ty.IsObjectType():
-		vals := make(map[string]cty.Value)
-		it := val.ElementIterator()
-		for it.Next() {
-			k, v := it.Element()
-			newVal := omitUnknowns(v)
-			if newVal != cty.NilVal {
-				vals[k.AsString()] = newVal
-			}
-		}
-		// We use object types always here, because the work we did above
-		// may have caused the individual elements to have different types,
-		// and we're doing this work to produce JSON anyway and JSON marshalling
-		// represents both of these mapping types as an object.
-		return cty.ObjectVal(vals)
-	default:
-		// Should never happen, since the above should cover all types
-		panic(fmt.Sprintf("omitUnknowns cannot handle %#v", val))
-	}
 }

--- a/internal/genconfig/generate_config_test.go
+++ b/internal/genconfig/generate_config_test.go
@@ -12,6 +12,7 @@ import (
 
 	"github.com/hashicorp/terraform/internal/addrs"
 	"github.com/hashicorp/terraform/internal/configs/configschema"
+	"github.com/hashicorp/terraform/internal/lang/marks"
 )
 
 func TestConfigGeneration(t *testing.T) {
@@ -538,6 +539,67 @@ resource "tfcoremock_simple_resource" "empty" {
   value = "[\"Hello\", \"World\""
 }`,
 		},
+		// Just try all the simple values with sensitive marks.
+		"sensitive_values": {
+			schema: &configschema.Block{
+				Attributes: map[string]*configschema.Attribute{
+					"string":       sensitiveAttribute(cty.String),
+					"empty_string": sensitiveAttribute(cty.String),
+					"number":       sensitiveAttribute(cty.Number),
+					"bool":         sensitiveAttribute(cty.Bool),
+					"object": sensitiveAttribute(cty.Object(map[string]cty.Type{
+						"nested": cty.String,
+					})),
+					"list": sensitiveAttribute(cty.List(cty.String)),
+					"map":  sensitiveAttribute(cty.Map(cty.String)),
+					"set":  sensitiveAttribute(cty.Set(cty.String)),
+				},
+			},
+			addr: addrs.AbsResourceInstance{
+				Module: addrs.RootModuleInstance,
+				Resource: addrs.ResourceInstance{
+					Resource: addrs.Resource{
+						Mode: addrs.ManagedResourceMode,
+						Type: "tfcoremock_sensitive_values",
+						Name: "values",
+					},
+					Key: addrs.NoKey,
+				},
+			},
+			provider: addrs.LocalProviderConfig{
+				LocalName: "tfcoremock",
+			},
+			value: cty.ObjectVal(map[string]cty.Value{
+				// Values that are sensitive will now be marked as such
+				"string":       cty.StringVal("Hello, world!").Mark(marks.Sensitive),
+				"empty_string": cty.StringVal("").Mark(marks.Sensitive),
+				"number":       cty.NumberIntVal(42).Mark(marks.Sensitive),
+				"bool":         cty.True.Mark(marks.Sensitive),
+				"object": cty.ObjectVal(map[string]cty.Value{
+					"nested": cty.StringVal("Hello, solar system!"),
+				}).Mark(marks.Sensitive),
+				"list": cty.ListVal([]cty.Value{
+					cty.StringVal("Hello, world!"),
+				}).Mark(marks.Sensitive),
+				"map": cty.MapVal(map[string]cty.Value{
+					"key": cty.StringVal("Hello, world!"),
+				}).Mark(marks.Sensitive),
+				"set": cty.SetVal([]cty.Value{
+					cty.StringVal("Hello, world!"),
+				}).Mark(marks.Sensitive),
+			}),
+			expected: `
+resource "tfcoremock_sensitive_values" "values" {
+  bool         = null # sensitive
+  empty_string = null # sensitive
+  list         = null # sensitive
+  map          = null # sensitive
+  number       = null # sensitive
+  object       = null # sensitive
+  set          = null # sensitive
+  string       = null # sensitive
+}`,
+		},
 	}
 	for name, tc := range tcs {
 		t.Run(name, func(t *testing.T) {
@@ -556,5 +618,13 @@ resource "tfcoremock_simple_resource" "empty" {
 				t.Errorf("got:\n%s\nwant:\n%s\ndiff:\n%s", got, want, diff)
 			}
 		})
+	}
+}
+
+func sensitiveAttribute(t cty.Type) *configschema.Attribute {
+	return &configschema.Attribute{
+		Type:      t,
+		Optional:  true,
+		Sensitive: true,
 	}
 }

--- a/internal/terraform/context_plan_import_test.go
+++ b/internal/terraform/context_plan_import_test.go
@@ -15,6 +15,7 @@ import (
 	"github.com/hashicorp/terraform/internal/configs/configschema"
 	"github.com/hashicorp/terraform/internal/plans"
 	"github.com/hashicorp/terraform/internal/providers"
+	testing_provider "github.com/hashicorp/terraform/internal/providers/testing"
 	"github.com/hashicorp/terraform/internal/states"
 	"github.com/hashicorp/terraform/internal/tfdiags"
 )
@@ -1413,7 +1414,7 @@ func TestContext2Plan_importGenerateNone(t *testing.T) {
 import {
   for_each = []
   to   = test_object.a
-  id   = "123"
+  id   = "81ba7c97"
 }
 `,
 	})
@@ -1435,5 +1436,68 @@ import {
 
 	if len(plan.Changes.Resources) != 0 {
 		t.Fatal("expected no resource changes")
+	}
+}
+
+// This is a test for the issue raised in #34992
+func TestContext2Plan_importWithSensitives(t *testing.T) {
+	m := testModuleInline(t, map[string]string{
+		"main.tf": `
+import {
+  to = test_object.a
+  id = "123"
+}
+`,
+	})
+
+	p := &testing_provider.MockProvider{
+		GetProviderSchemaResponse: &providers.GetProviderSchemaResponse{
+			ResourceTypes: map[string]providers.Schema{
+				"test_object": {
+					Block: &configschema.Block{
+						Attributes: map[string]*configschema.Attribute{
+							"sensitive_string": {
+								Type:      cty.String,
+								Sensitive: true,
+								Optional:  true,
+							},
+							"sensitive_list": {
+								Type:      cty.List(cty.String),
+								Sensitive: true,
+								Optional:  true,
+							},
+						},
+					},
+				},
+			},
+		},
+		ImportResourceStateFn: func(request providers.ImportResourceStateRequest) providers.ImportResourceStateResponse {
+			return providers.ImportResourceStateResponse{
+				ImportedResources: []providers.ImportedResource{
+					{
+						TypeName: "test_object",
+						State: cty.ObjectVal(map[string]cty.Value{
+							"sensitive_string": cty.StringVal("sensitive"),
+							"sensitive_list":   cty.ListVal([]cty.Value{cty.StringVal("hello"), cty.StringVal("world")}),
+						}),
+					},
+				},
+			}
+		},
+	}
+
+	ctx := testContext2(t, &ContextOpts{
+		Providers: map[addrs.Provider]providers.Factory{
+			addrs.NewDefaultProvider("test"): testProviderFuncFixed(p),
+		},
+	})
+
+	// Just don't crash!
+	_, diags := ctx.Plan(m, states.NewState(), &PlanOpts{
+		Mode:               plans.NormalMode,
+		GenerateConfigPath: "generated.tf",
+	})
+	if diags.HasErrors() {
+		t.Fatalf("unexpected errors\n%s", diags.Err().Error())
 	}
 }

